--- kmod-17.orig/libkmod/libkmod-internal.h	2014-04-01 05:40:37.158606756 -0500
+++ kmod-17/libkmod/libkmod-internal.h	2015-04-27 12:18:20.433948434 -0500
@@ -45,6 +45,10 @@
 #  endif
 #endif
 
+#ifndef O_CLOEXEC
+# define O_CLOEXEC 0
+#endif
+
 void kmod_log(const struct kmod_ctx *ctx,
 		int priority, const char *file, int line, const char *fn,
 		const char *format, ...) __attribute__((format(printf, 6, 7))) __attribute__((nonnull(1, 3, 5)));
